ipapython/ccache_storage.py

```diff
@@ -0,0 +1,234 @@
+
+    krb5_cc_set_config = LIBKRB5.krb5_cc_set_config
+    krb5_cc_set_config.argtypes = (krb5_context, krb5_ccache, krb5_principal,
+                                   ctypes.c_char_p, krb5_data_p, )
+    krb5_cc_set_config.retval = krb5_error
+    krb5_cc_set_config.errcheck = krb5_errcheck
+
+    krb5_cc_get_config = LIBKRB5.krb5_cc_get_config
+    krb5_cc_get_config.argtypes = (krb5_context, krb5_ccache, krb5_principal,
+                                   ctypes.c_char_p, krb5_data_p, )
+    krb5_cc_get_config.retval = krb5_error
+    krb5_cc_get_config.errcheck = krb5_errcheck
+
+class session_store:
+    def __init__(self, name='X-IPA-Session-Cookie'):
+        self.__context = None
+        if isinstance(LIBKRB5, Exception):  # pragma: no cover
+            raise LIBKRB5
+        context = krb5_context()
+        krb5_init_context(ctypes.byref(context))
+        self.__context = context
+
+        self._hidden_cred_name = name
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self.__context:
+            krb5_free_context(self.__context)
+            self.__context = None
+
+    def __del__(self):
+        self.__exit__(None, None, None)
+
+    def store_data(self, client, value):
+        """
+        Stores the session cookie in a hidden ccache entry.
+        """
+        assert isinstance(client, six.string_types)
+        assert isinstance(value, bytes)
+
+        principal = ccache = None
+
+        try:
+            principal = krb5_principal()
+            krb5_parse_name(self.__context, ctypes.c_char_p(client),
+                            ctypes.byref(principal))
+
+            ccache = krb5_ccache()
+            krb5_cc_default(self.__context, ctypes.byref(ccache))
+
+            buf = ctypes.create_string_buffer(value)
+            data = _krb5_data()
+            data.data = buf.value
+            data.length = len(buf)
+            krb5_cc_set_config(self.__context, ccache, principal,
+                               self._hidden_cred_name, ctypes.byref(data))
+
+        finally:
+            if principal:
+                krb5_free_principal(self.__context, principal)
+            if ccache:
+                krb5_cc_close(self.__context, ccache)
+
+    def get_data(self, client):
+        """
+        Gets the session cookie in a hidden ccache entry.
+        """
+        assert isinstance(client, six.string_types)
+
+        principal = ccache = data = None
+
+        try:
+            principal = krb5_principal()
+            krb5_parse_name(self.__context, ctypes.c_char_p(client),
+                            ctypes.byref(principal))
+
+            ccache = krb5_ccache()
+            krb5_cc_default(self.__context, ctypes.byref(ccache))
+
+            data = _krb5_data()
+            krb5_cc_get_config(self.__context, ccache, principal,
+                               self._hidden_cred_name, ctypes.byref(data))
+
+            return str(data.data)
```

Comment:
What is `data.data` type? If bytes this will return unexpected string  in py3. `.decode()` should be used in case of bytes